src/kernel/thread.c

```diff
@@ -307,6 +307,10 @@ static void nextDomain(void)
     }
 #ifdef CONFIG_KERNEL_MCS
     NODE_STATE(ksReprogram) = true;
+#endif
+#ifdef CONFIG_HAVE_FPU
+    /* Save FPU state now to avoid touching cross-domain state later: */
+    switchLocalFpuOwner(NULL);
 #endif
```

Comment:
I've been working on something similar recently, which involved flushing the vcpu before switching domain. Based on that, it helps verification if we can keep `nextDomain`  so that it only modifies the domain related parts of the state, and separate out the other state updates that we're adding. 

How would you feel about something like a `prepareNextDomain` , in the same way that I've done in https://github.com/seL4/seL4/compare/master...corlewis:vcpu_flush.

Reply:
Thanks, that's good to know for x86, I hadn't looked into other architectures at all yet.

Probably best not to merge it, I've updated the proofs for that change on aarch64 but there's a related issue that's causing problems and is going to take a while to work through. If you're ok with that structure then adding `Arch_prepareNextDomain` with the FPU clearing would be good, and then once the VCPU work is finished they can both be cleared in the same function.

Reply:
I've added a `prepareNextDomain`, as it seems more config dependent than arch dependent. We can change it later if you want.

Reply:
That's fine, I think I went with the arch naming because I came at it from the verification perspective, where we treat different configs like this very similarly to architectures.